pkg/server/profiles.go

```diff
@@ -0,0 +1,187 @@
+	if err != nil {
+		return nil, fmt.Errorf("could not create client config: %w", err)
+	}
+
+	_, rawClient, err := kube.NewKubeHTTPClientWithConfig(rest, clusterName)
+	if err != nil {
+		return nil, fmt.Errorf("could not create kube http client: %w", err)
+	}
+
+	tempDir, err := ioutil.TempDir("", "helmrepocache")
+	if err != nil {
+		return nil, err
+	}
+
+	profilesSrv := &ProfilesServer{
+		KubeClient:        rawClient,
+		Log:               logr,
+		HelmChartManager:  helm.NewRepoManager(rawClient, tempDir),
+		HelmRepoNamespace: helmRepoNamespace,
+		HelmRepoName:      helmRepoName,
+		cacheDir:          tempDir,
+	}
+
+	mux := runtime.NewServeMux(middleware.WithGrpcErrorLogging(logr))
+	httpHandler := middleware.WithLogging(logr, mux)
+
+	if err := pb.RegisterProfilesHandlerServer(ctx, mux, profilesSrv); err != nil {
+		return nil, fmt.Errorf("could not register Profiles server: %w", err)
+	}
+
+	return httpHandler, nil
+}
+
+type ProfilesServer struct {
+	pb.UnimplementedProfilesServer
+
+	KubeClient        client.Client
+	Log               logr.Logger
+	HelmChartManager  HelmRepoManager
+	HelmRepoName      string
+	HelmRepoNamespace string
+	cacheDir          string
+}
+
+func (s *ProfilesServer) GetProfiles(ctx context.Context, msg *pb.GetProfilesRequest) (*pb.GetProfilesResponse, error) {
+	helmRepo := &sourcev1beta1.HelmRepository{}
+	err := s.KubeClient.Get(ctx, client.ObjectKey{
+		Name:      s.HelmRepoName,
+		Namespace: s.HelmRepoNamespace,
+	}, helmRepo)
+
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			errMsg := fmt.Sprintf("HelmRepository %q/%q does not exist", s.HelmRepoNamespace, s.HelmRepoName)
+			s.Log.Error(err, errMsg)
+
+			return &pb.GetProfilesResponse{
+					Profiles: []*pb.Profile{},
+				}, &grpcruntime.HTTPStatusError{
+					Err:        errors.New(errMsg),
+					HTTPStatus: http.StatusOK,
+				}
+		}
+
+		return nil, fmt.Errorf("failed to get HelmRepository %q/%q", s.HelmRepoNamespace, s.HelmRepoName)
+	}
+
+	ps, err := s.HelmChartManager.GetCharts(ctx, helmRepo, helm.Profiles)
+	if err != nil {
+		return nil, fmt.Errorf("failed to scan HelmRepository %q/%q for charts: %w", s.HelmRepoNamespace, s.HelmRepoName, err)
+	}
+
+	return &pb.GetProfilesResponse{
+		Profiles: ps,
+	}, nil
+}
+
+func (s *ProfilesServer) GetProfileValues(ctx context.Context, msg *pb.GetProfileValuesRequest) (*httpbody.HttpBody, error) {
+	helmRepo := &sourcev1beta1.HelmRepository{}
+	err := s.KubeClient.Get(ctx, client.ObjectKey{
+		Name:      s.HelmRepoName,
+		Namespace: s.HelmRepoNamespace,
+	}, helmRepo)
+
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			errMsg := fmt.Sprintf("HelmRepository %q/%q does not exist", s.HelmRepoNamespace, s.HelmRepoName)
+			s.Log.Error(err, errMsg)
+
+			return &httpbody.HttpBody{
+					ContentType: "application/json",
+					Data:        []byte{},
+				}, &grpcruntime.HTTPStatusError{
+					Err:        errors.New(errMsg),
+					HTTPStatus: http.StatusOK,
```

Comment:
shouldn't this be `http.StatusNotFound` or `http. StatusBadRequest` ? 🤔

Reply:
We thought similar when we first looked at this, it appears to be a very intentional - https://github.com/weaveworks/weave-gitops-enterprise/pull/245/commits/aaccb2dc4183bef9c7aee3ba391077c931d1be90

I want to change it as it feels wrong to me, but I think we need to change the behaviour in the UI first to handle this change gracefully before making the change in the API.

Reply:
yeah, we should definitely change this 👍🏻  why would we wait for tho, this is in core and the UI for EE is not using this. The graceful handling of UI can be part of https://github.com/weaveworks/weave-gitops-enterprise/issues/311  when we switch over? 🤔

Reply:
I sort of assumed we wouldn't be making the UI changes (I have no idea even how to :smile: ). So we would need to raise it with them, let them make the changes (e.g. when a 404 is returned show an empty list if thats the behaviour they want) and then change it in core

Reply:
if your happy with ^ I'll create an issue in EE and raise it with them

Reply:
yeah, I didn't mean we should do the UI changes, let's open a ticket in EE for UI to handle this change 👍🏻

Reply:
https://github.com/weaveworks/weave-gitops-enterprise/issues/366